commit 21a4cc00e8e67edcfc1bdb9af6d370ed1226eb86
Author: John W. Linville <linville@tuxdriver.com>
Date:   Thu May 28 11:39:02 2009 +0200

    at76c50x-usb: avoid mutex deadlock in at76_dwork_hw_scan
    
    http://bugzilla.kernel.org/show_bug.cgi?id=13312
    
    at76_dwork_hw_scan holds a mutex while calling ieee80211_scan_completed,
    which then calls at76_config which needs the same mutex.  This reworks
    the ordering to not hold the lock while calling ieee80211_scan_completed.
    
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

--- a/drivers/net/wireless/at76c50x-usb.c
+++ b/drivers/net/wireless/at76c50x-usb.c
@@ -1873,18 +1873,18 @@ static void at76_dwork_hw_scan(struct wo
 	if (ret != CMD_STATUS_COMPLETE) {
 		queue_delayed_work(priv->hw->workqueue, &priv->dwork_hw_scan,
 				   SCAN_POLL_INTERVAL);
-		goto exit;
+		mutex_unlock(&priv->mtx);
+		return;
 	}
 
-	ieee80211_scan_completed(priv->hw);
-
 	if (is_valid_ether_addr(priv->bssid))
 		at76_join(priv);
 
-	ieee80211_wake_queues(priv->hw);
-
-exit:
 	mutex_unlock(&priv->mtx);
+
+	ieee80211_scan_completed(priv->hw);
+
+	ieee80211_wake_queues(priv->hw);
 }
 
 static int at76_hw_scan(struct ieee80211_hw *hw, u8 *ssid, size_t len)
